vertexai/genai/tokenizer/tokenizer.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tokenizer provides local token counting for Gemini models. This
+// tokenizer downloads its model from the web, but otherwise doesn't require
+// an API call for every CountTokens invocation.
+package tokenizer
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"cloud.google.com/go/vertexai/genai"
+	"cloud.google.com/go/vertexai/internal/sentencepiece"
+)
+
+var supportedModels = map[string]bool{
+	"gemini-1.0-pro":       true,
+	"gemini-1.5-pro":       true,
+	"gemini-1.5-flash":     true,
+	"gemini-1.0-pro-001":   true,
+	"gemini-1.0-pro-002":   true,
+	"gemini-1.5-pro-001":   true,
+	"gemini-1.5-flash-001": true,
+}
+
+// Tokenizer is a local tokenizer for text.
+type Tokenizer struct {
+	encoder *sentencepiece.Encoder
+}
+
+// CountTokensResponse is the response of [Tokenizer.CountTokens].
+type CountTokensResponse struct {
+	TotalTokens int32
+}
+
+// New creates a new [Tokenizer] from a model name; the model name is the same
+// as you would pass to a [genai.Client.GenerativeModel].
+func New(modelName string) (*Tokenizer, error) {
+	if !supportedModels[modelName] {
+		return nil, fmt.Errorf("model %s is not supported", modelName)
+	}
+
+	data, err := loadModelData(gemmaModelURL, gemmaModelHash)
+	if err != nil {
+		return nil, fmt.Errorf("loading model: %w", err)
+	}
+
+	encoder, err := sentencepiece.NewEncoder(bytes.NewReader(data))
+	if err != nil {
+		return nil, fmt.Errorf("creating encoder: %w", err)
+	}
+
+	return &Tokenizer{encoder: encoder}, nil
+}
+
+// CountTokens counts the tokens in all the given parts and returns their
+// sum. Only [genai.Text] parts are suppored; an error will be returned if
```

Comment:
Then why not make the argument of type ...genai.Text?

Reply:
I was thinking it would be awkward for this to have a different type from https://pkg.go.dev/cloud.google.com/go/vertexai/genai#GenerativeModel.CountTokens

Consider wanting to count tokens with both. One would likely create a `[]Content`, because of the invariance issue we can't pass `[]Text` to `GenerativeModel.CountTokens`. But then to call `Tokenizer.CountTokens` one would then have to convert `[]Content` to `[]Text` which is awkward.

Also if in the future `Tokenizer.CountTokens` supports other parts, changing it to `[]Content` will be a breaking change.